api/lib/data-cube/project-source/create.ts

```diff
@@ -1,24 +1,28 @@
 import { Request, Response } from 'express'
 import parse from 'csv-parse'
-import detect from 'detect-csv'
 import asyncMiddleware from 'middleware-async'
 import { createSource } from '../../domain/project'
 import { projects, sources } from '../../storage/repository'
 import { NotFoundError } from '../../error'
 import { getProjectId } from '../../read-graphs/project/links'
 import { getRepresentation } from '../../read-graphs/source/index'
 import env from '../../env'
+import CSVSniffer = require('csv-sniffer')
+
+const csvDelimiters = [',', ';', '\t']
+
+const sniffer = new (CSVSniffer())(csvDelimiters)
 
 export function parseCsv (req, res, next) {
-  let delimiter = ','
-  const detectedCsvFormat = detect(req.body)
-  if (detectedCsvFormat) {
-    delimiter = detectedCsvFormat.delimiter
-  }
+  res.locals.csvDialect = { }
+
+  const detectedCsvFormat = sniffer.sniff(req.body)
+  res.locals.csvDialect.delimiter = detectedCsvFormat.delimiter
+  res.locals.csvDialect.quote = detectedCsvFormat.quoteChar || ''
```

Comment:
Shouldn't it be `null` instead of `''` here? (I don't understand the flow of data in the API, so I'm really not sure. Just a hunch).

Reply:
I don't know if it makes a difference to the csvw parser but I'd rather simply avoid null

Reply:
And it won't cause issues with the `valueOrDefault` function you defined somewhere else?

Reply:
No, empty string is a valid value AFAIK. It will not be replaced by `valueOrDefault`

Reply:
Yeah but that's what concerns me. If the "sniffer" returns `null`, does it mean that the quote char could not be identified or that there are none? If it could not be identified, maybe we should use our default?

Reply:
The docs say 'null or empty string means no quote character' so here's where this came from

Reply:
It says that for the "options" of the function. For the returned object, it only says "quoteChar [string]: Can be either ' or " or null.". That's why I was wondering.

For example, I have no idea what happens if there are no escaped strings in the first 100 lines, but there are some in the rest of the file. I guess it's an impossible problem to solve... let's go with your solution :)

Reply:
Oh, my bad. But yea, if there are quotes somewhere beyond the sample, you will not know anyway. Indeed, maybe double quote could be more common and thus a safer fallback

Reply:
Yeah, maybe. It depends on the behavior of the parser if you tell it that `"` is the delimiter when it's not...

Reply:
Might case issues. I think it might fail completely is the header is quoted like `'station id'` but you tell the parser the char is `"` then the column name will be wrong and the entire mapping  falls apart :)

But any wrong guess will cause that then

Reply:
I think in the case you described, the sniffer will pick up `'` as the quoting char. I was thinking more if you tell the parser that `"` is the quoting char, and then `"` is used in string values, but not as a quoting char. Anyway, I don't think we can solve this issue magically. Let's hope it's rare enough ;)

Reply:
I see. Logically, I think the delimiter is only considered when directly adjacent to the delimiter. That is also its purpose: to allow the delimiter inside the cell value

```
cell 1, cell 2,"cell, with, commas"
```

Reply:
Yeah that makes sense